### ISSUE TITLE
Errors

### ISSUE THREAD

**Issue by WojtekMs** [Vicx95/SharedPtr#1]
https://github.com/Vicx95/SharedPtr/blob/e848754bfca2592a57292bb67afa4d852428b88f/sharedptr.hpp#L29
You call delete to object that you didn't allocate memory for! This is causing segmentation fault in case I create an object on the stack and if I create object on heap you free memory of my object without my knowledge. This is very bad.
Code like this causes error:
```
int random = 5;
SharedPtr<int> s_ptr(&random);
```
Because you cannot call delete for an object / variable allocated on stack
Code like this causes error:
```
int* random = new int{5};
SharedPtr<int> s_ptr(random);
delete random;
```
Because you cannot call delete twice for the same object / variable

**Comment by Vicx95** [comment]
@WojtekMs Yes but, if you use shared_ptr from std library it's the same error :) 
`int random = 5;
    std::shared_ptr<int> s_ptr(&random);

free(): invalid pointer
Aborted (core dumped)

 int* random = new int(5);
    std::shared_ptr<int> s_ptr(random);
    delete random;
free(): double free detected in tcache 2
Aborted (core dumped)
`

**Comment by WojtekMs** [comment]
https://github.com/Vicx95/SharedPtr/blob/e848754bfca2592a57292bb67afa4d852428b88f/sharedptr.hpp#L17
In this move constructor there are memory leaks.
In the code bloew pointer refCounter_ is being overwrriten yet it has been allocated on the heap in constructor! This way we loose our pointer to this address, which was not freed and computer cannot use it again until the program ends.
`refCounter_ = sharedPtr.refCounter_;`
Another memory leak in the same move constructor is here:
`sharedPtr.refCounter_ = new RefCounter(); `
Temporary's object member refCounter has random value! And in destructor you check if this random value == 0 and only if it's true you call delete.

**Comment by WojtekMs** [comment]
In case we use default c-tor (ptr is nullptr) refCounter is not being incremented, yet in the destructor refCounter is always decremented. If refCounter value is 0 (which is not certain because no value was explicitly given to it) in the destructor it will be decremented and the memory will not be freed.

**Comment by WojtekMs** [comment]
> 
> 
> @WojtekMs Yes but, if you use shared_ptr from std library it's the same error, because shared_ptr manage memory allocated on heap not stack
> `int random = 5;
> std::shared_ptr s_ptr(&random);
> 
> free(): invalid pointer
> Aborted (core dumped)
> 
> int* random = new int(5);
> std::shared_ptr s_ptr(random);
> delete random;
> free(): double free detected in tcache 2
> Aborted (core dumped)
> `

You are right. My bad

**Comment by Vicx95** [comment]
> In case we use default c-tor (ptr is nullptr) refCounter is not being incremented, yet in the destructor refCounter is always decremented. If refCounter value is 0 (which is not certain because no value was explicitly given to it) in the destructor it will be decremented and the memory will not be freed.

I'm repair it now :)

**Comment by Vicx95** [comment]
> https://github.com/Vicx95/SharedPtr/blob/e848754bfca2592a57292bb67afa4d852428b88f/sharedptr.hpp#L17
> 
> In this move constructor there are memory leaks.
> In the code below pointer refCounter_ is being overwrriten yet it has been allocated on the heap in constructor! This way we loose our pointer to this address, which was not freed and computer cannot use it again until the program ends.
> `refCounter_ = sharedPtr.refCounter_;`

Hmm, I'm using valgrind on this code and below you have a report from Valgrind
`SharedPtr<int> s_ptr(new int(5));
    SharedPtr<int> s_ptr2 = s_ptr;

    SharedPtr<int> s_ptr3(std::move(s_ptr));
algrind --leak-check=full ./sharedptr 
==10005== Memcheck, a memory error detector
==10005== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==10005== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==10005== Command: ./sharedptr
==10005== 
0SHARED PTR DESTRUCTOR
SHARED PTR DESTRUCTOR
==10005== 
==10005== HEAP SUMMARY:
==10005==     in use at exit: 0 bytes in 0 blocks
==10005==   total heap usage: 5 allocs, 5 frees, 73,740 bytes allocated
==10005== 
==10005== All heap blocks were freed -- no leaks are possible
==10005== 
==10005== For lists of detected and suppressed errors, rerun with: -s
==10005== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

`

**Comment by WojtekMs** [comment]
Hard to say for now... I will try to think about it tomorrow, but also for now I have this error: 
```
SharedPtr<int> s_ptr0;
SharedPtr<int> s_ptr(s_ptr0);
```
valgrind gives info:
==23672== HEAP SUMMARY:
==23672==     in use at exit: 4 bytes in 1 blocks
==23672==   total heap usage: 3 allocs, 2 frees, 73,732 bytes allocated
==23672== 
==23672== LEAK SUMMARY:
==23672==    definitely lost: 4 bytes in 1 blocks
==23672==    indirectly lost: 0 bytes in 0 blocks
==23672==      possibly lost: 0 bytes in 0 blocks
==23672==    still reachable: 0 bytes in 0 blocks
==23672==         suppressed: 0 bytes in 0 blocks
==23672== Rerun with --leak-check=full to see details of leaked memory
